= Extract TAIR accession IDs

== TAIR
TAIR (The Arabidopsis Information Resource) is a very useful website that contains information on the Arabidopsis genes and genome. IDs used by the website follow a naming convention. For example:


* `AT1G74310` (HEAT SHOCK PROTEIN 101, HOT1)
** `AT` 1G74310 means that the gene is in Arabidopsis thaliana
** AT `1` G74310 means that the gene is on chromosome 1
** AT1 `G74310` is the unique gene identifier


A number of program outputs will contain only these TAIR identifiers, for example RNAseq data. While it would be interesting to look at each gene in a data set and have a description of it immediately, it may not be the most useful for some data sets.
For example, I have an RNAseq data set with a list of TAIR IDs and their expression levels. I want to identify the expression pattern of a subset of these genes. In particular, I want to look at the expression of genes that typically are upregulated in response to a heat stress.
I can go to the TAIR website and find a list of genes that are associated with the identifier `response to heat`. However, after I download that list, it gives me all this extra info:

     TAIR Accession	Locus	Gene Model	Gene Type	Description	Other Name(Type)	Keywords	Is full length cDNA
    Gene:1005715766	AT5G67030	AT5G67030.2	protein_coding	Encodes a single copy zeaxanthin epoxidase gene that functions in first step of the biosynthesis of the abiotic stress hormone abscisic acid (ABA). Mutants in this gene are unable to express female sterility in response to beta-aminobutyric acid, as wild type plants do. ABA DEFICIENT 1; ABA1; ARABIDOPSIS THALIANA ABA DEFICIENT 1; ARABIDOPSIS THALIANA ZEAXANTHIN EPOXIDASE; ATABA1; ATZEP; IBS3; IMPAIRED IN BABA-INDUCED STERILITY 3; LOS6; LOW EXPRESSION OF OSMOTIC STRESS-RESPONSIVE GENES 6; NON-PHOTOCHEMICAL QUENCHING 2; NPQ2; ZEAXANTHIN EPOXIDASE; ZEP chloroplast, chloroplast envelope, FAD binding, LP.02 two leaves visible stage, LP.04 four leaves visible stage, LP.06 six leaves visible stage, LP.08 eight leaves visible stage, LP.10 ten leaves visible stage, LP.12 twelve leaves visible stage, abscisic acid biosynthetic process, carpel, cauline leaf, chloroplast, chloroplast envelope, collective leaf structure, cotyledon, flower, flower pedicel, flowering stage, guard cell, hypocotyl, inflorescence meristem, leaf apex, leaf lamina base, mature plant embryo stage, membrane, petal, petal differentiation and expansion stage, petiole, plant embryo, plant embryo bilateral stage, plant embryo cotyledonary stage, plant embryo globular stage, plastid, pollen, response to heat, response to osmotic stress, response to red light, response to water deprivation, root, rosette leaf, seed, sepal, shoot apex, shoot system, stamen, stem, sugar mediated signaling pathway, vascular leaf, vascular leaf senescent stage, xanthophyll biosynthetic process, zeaxanthin epoxidase [overall] activity	true

Super interesting... But all I wanted was the list of TAIR accession numbers!

Alright, maybe I can obtain the data another way. There is a general resource website for such identifiers as the `response to heat`. These identifiers are also sometimes called GO terms, and you can look up these terms on `http://amigo.geneontology.org`. But even there, when I download the file, I get all this extra information I don't want or need:

    AT5G12020|17.6 kDa class II heat shock protein|F14F18.190|F14F18_190
    AT5G12020|17.6 kDa class II heat shock protein|F14F18.190|F14F18_190
    AT5G41340|ATUBC4|ubiquitin conjugating enzyme 4|MYC6.5|MYC6_5

Should I copy and paste only the TAIR accession numbers out of these lists by hand? It should be simple enough when I have two or three, but when my list contains over 300 IDs? I think not!

Solution: write a program to do it for you!

== The Program
Write a python program called `accession.py` that will extract TAIR accession IDs from a file and write those IDs to an outfile.